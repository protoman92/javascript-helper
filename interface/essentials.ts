import { DeepPartial } from "ts-essentials";
import { GenericAsyncFunction, GenericFunction } from ".";

export type ArrayOrSingle<T> = T | T[];

export type ArrayToMap<
  T extends any[] | readonly any[],
  TKey extends keyof T[number]
> = Map<T[number][TKey], T[number]>;

export type Promised<T> = T extends Promise<infer U> ? U : T;
export type Mapper<T1, T2 = T1> = (args: T1) => T2;
export type MapEntry<M> = M extends Map<infer K, infer V> ? [K, V] : never;
export type MapKey<M> = M extends Map<infer K, unknown> ? K : never;

export type MapFromObject<O> = O extends Record<infer K, infer V>
  ? Map<K, V>
  : never;

export type MapToObject<M> = M extends Map<infer K, infer V>
  ? Record<Extract<K, number | string | symbol>, V>
  : never;

export type MapValue<M> = M extends Map<unknown, infer V> ? V : never;

export type NonNullableProps<
  A extends Record<number | string | symbol, any>,
  K extends keyof A = keyof A
> = Required<{ [x in K]: NonNullable<A[x]> }> & Omit<A, K>;

export type Nullable<T> = T | null;
export type NullableProps<T> = { [x in keyof T]: Nullable<T[x]> };
export type Neverable<T> = T | null | undefined;
export type ObjectValue<O> = O extends Record<any, infer V> ? V : never;

export type Promisified<
  FN extends GenericFunction
> = FN extends GenericAsyncFunction
  ? FN
  : (...args: Parameters<FN>) => Promise<ReturnType<FN>>;

export type Returnable<T, Args extends unknown[] = []> =
  | T
  | ((...args: Args) => T);

export type Resolvable<T> = T | Promise<T> | PromiseLike<T>;
export type SetToArray<T> = T extends Set<infer V> ? V[] : never;
export type SetValue<T> = T extends Set<infer V> ? V : never;

export type WithDeepPartialReturn<FN extends (...args: any[]) => any> = (
  ...args: Parameters<FN>
) => ReturnType<FN> extends Promise<infer T>
  ? Promise<DeepPartial<T>>
  : DeepPartial<ReturnType<FN>>;
