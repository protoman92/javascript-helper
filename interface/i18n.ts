import { GenericObject as O } from ".";

export namespace I18N {
  export type LocaleContent = { [x: string]: string | LocaleContent };
  export type TemplateReplacement = Record<string, unknown>;
}

/**
 * Represents an i18n client that has a type-checked translate method. The
 * types will force keys to be entered until they resolve to a string.
 * The level of nest depends on the translation object. If necessary, we can
 * add more nested key types.
 */
export interface I18N<T extends I18N.LocaleContent> {
  t<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    K4 extends keyof T[K1][K2][K3]
  >(
    ...keys: [
      ...(T extends O
        ? [
            K1,
            ...(T[K1] extends O
              ? [
                  K2,
                  ...(T[K1][K2] extends O
                    ? [
                        K3,
                        ...(T[K1][K2][K3] extends O
                          ? [K4]
                          : Partial<[I18N.TemplateReplacement]>)
                      ]
                    : Partial<[I18N.TemplateReplacement]>)
                ]
              : Partial<[I18N.TemplateReplacement]>)
          ]
        : Partial<[I18N.TemplateReplacement]>)
    ]
  ): string;

  t<
    K1 extends Extract<keyof T, string>,
    K2 extends Extract<keyof T[K1], string>,
    K3 extends Extract<keyof T[K1][K2], string>,
    K4 extends Extract<keyof T[K1][K2][K3], string>
  >(
    key: T extends O
      ? `${K1}${T[K1] extends O
          ? `.${K2}${T[K1][K2] extends O
              ? `.${K3}${T[K1][K2][K3] extends O ? `.${K4}` : ""}`
              : ""}`
          : ""}`
      : "",
    replacement?: I18N.TemplateReplacement
  ): string;

  t<K extends string>(
    rawKey: K extends keyof T ? never : K,
    replacement?: I18N.TemplateReplacement
  ): string;
}
