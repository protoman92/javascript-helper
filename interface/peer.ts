import Peer, { DataConnection } from "peerjs";
import { Observable } from "rxjs";
import { EventEmitterClient, LifecycleAware, Mappable, MessageSender } from ".";
export type PeerConnection = DataConnection;

declare namespace PeerEvent {
  interface BaseEvent {
    readonly peerID: string;
  }

  export interface Data<T> extends BaseEvent {
    readonly data: T;
    readonly type: "DATA";
  }

  export interface Open extends BaseEvent {
    readonly type: "OPEN";
  }
}

export type PeerEvent<T> = PeerEvent.Data<T> | PeerEvent.Open;

export namespace PeerState {
  export interface PeerError {
    readonly error: Error;
    readonly type: "ERROR";
  }

  export interface PeerOpen {
    readonly peerID: string;
    readonly type: "OPEN";
  }
}

export type PeerState = PeerState.PeerError | PeerState.PeerOpen;

export interface PeerClient extends LifecycleAware, Mappable {
  connectToPeer(
    ...args: Parameters<Peer["connect"]>
  ): Observable<PeerConnection>;

  onPeerConnection(): Observable<PeerConnection>;
  peerState$(): Observable<PeerState | undefined>;
  streamPeerEvents<T>(conn: PeerConnection): Observable<PeerEvent<T>>;
}

export type PeerClientFactory = ReturnType<
  typeof import("../client/peer_client_factory")["default"]
>;

export namespace ManyToOnePeerClient {
  interface ConnectionOpenArgs {
    readonly connection: PeerConnection;
  }

  interface DataReceiveArgs<SubscribeeData> {
    readonly data: SubscribeeData;
  }

  namespace OutgoingConnectionUpdateArgs {
    interface BaseArgs {
      readonly allPeers: Readonly<{ id: string }>[];
    }

    export interface PeerJoining extends BaseArgs {
      readonly joiningPeerID: string;
      readonly type: "PEER_JOINING";
    }

    export interface PeerLeaving extends BaseArgs {
      readonly leavingPeerID: string;
      readonly type: "PEER_LEAVING";
    }
  }

  type OutgoingConnectionUpdateArgs =
    | OutgoingConnectionUpdateArgs.PeerJoining
    | OutgoingConnectionUpdateArgs.PeerLeaving;

  interface RetryElapseArgs {
    readonly elapsed: number;
  }

  export type Callbacks<SubscribeeData> = Readonly<{
    connectionOpen: (args: ConnectionOpenArgs) => void;
    dataReceive: (args: DataReceiveArgs<SubscribeeData>) => void;
    outgoingConnectionUpdate: (args: OutgoingConnectionUpdateArgs) => void;
    retryCancel: () => void;
    retryElapse: (args: RetryElapseArgs) => void;
  }>;
}

export interface ManyToOnePeerClient<SubscribeeData>
  extends LifecycleAware,
    Pick<
      EventEmitterClient<ManyToOnePeerClient.Callbacks<SubscribeeData>>,
      "callbacksType" | "off" | "on"
    >,
    MessageSender<SubscribeeData> {}
