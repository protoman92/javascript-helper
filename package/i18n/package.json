{
  "author": "Hai Pham <swiften.svc@gmail.com",
  "dependencies": {
    "@haipham/javascript-helper-essential-types": "workspace:*"
  },
  "devDependencies": {
    "@types/jest": "^29.4.0",
    "babel-jest": "^29.4.3",
    "jest": "^29.4.3",
    "prettier": "^2.7.1",
    "sort-json": "^2.0.1",
    "ts-essentials": "^9.3.1",
    "ts-jest": "^29.0.5",
    "typescript": "^4.9.3"
  },
  "license": "MIT",
  "main": "src/index.ts",
  "name": "@haipham/javascript-helper-i18n",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "echo \"Skipped\"",
    "test": "jest"
  },
  "version": "1.0.1"
}
