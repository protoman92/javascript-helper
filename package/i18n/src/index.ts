import { GenericObject as O } from "@haipham/javascript-helper-essential-types";

export namespace I18N {
  export type LocaleContent = { [x: string]: string | LocaleContent };

  export type Options<TOptions = {}> = Readonly<{
    replacement?: TemplateReplacement;
  }> &
    TOptions;

  export type TemplateReplacement = Record<string, unknown>;
}

type KeyArray<
  T,
  K1 extends keyof T,
  K2 extends keyof T[K1],
  K3 extends keyof T[K1][K2],
  K4 extends keyof T[K1][K2][K3]
> = [
  ...(T extends O
    ? [
        K1,
        ...(T[K1] extends O
          ? [
              K2,
              ...(T[K1][K2] extends O
                ? [K3, ...(T[K1][K2][K3] extends O ? [K4] : [])]
                : [])
            ]
          : [])
      ]
    : [])
];

/**
 * Represents an i18n client that has a type-checked translate method. The
 * types will force keys to be entered until they resolve to a string.
 * The level of nest depends on the translation object. If necessary, we can
 * add more nested key types.
 * @template TOptions additional options that can be specified to work with
 * different underlying i18n implementations.
 */
export interface I18N<T extends I18N.LocaleContent, TOptions = {}> {
  k<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    K4 extends keyof T[K1][K2][K3]
  >(
    ...keys: KeyArray<T, K1, K2, K3, K4>
  ): typeof keys;

  t<
    K1 extends keyof T,
    K2 extends keyof T[K1],
    K3 extends keyof T[K1][K2],
    K4 extends keyof T[K1][K2][K3]
  >(
    ...keys: [
      ...KeyArray<T, K1, K2, K3, K4>,
      ...Partial<[I18N.Options<TOptions>]>
    ]
  ): string;

  t<K extends string>(
    rawKey: K extends keyof T ? never : K,
    options?: I18N.Options<TOptions>
  ): string;
}
