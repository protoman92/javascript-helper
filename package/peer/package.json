{
  "author": "Hai Pham <swiften.svc@gmail.com",
  "dependencies": {
    "@haipham/javascript-helper-event-emitter": "workspace:*",
    "@haipham/javascript-helper-object": "workspace:*",
    "@haipham/javascript-helper-rxjs": "workspace:*",
    "lodash-es": "^4.17.21",
    "peerjs": "^1.4.7",
    "rxjs": "^6.6.3"
  },
  "devDependencies": {
    "@haipham/javascript-helper-client-types": "workspace:*",
    "@types/jest": "^29.4.0",
    "@types/lodash-es": "^4.17.6",
    "babel-jest": "^29.4.3",
    "jest": "^29.4.3",
    "jest-environment-jsdom": "~29.4.3",
    "sort-json": "^2.0.1",
    "ts-essentials": "^9.0.0",
    "ts-jest": "^29.0.5",
    "typescript": "^4.9.3"
  },
  "license": "MIT",
  "main": "src/index.ts",
  "name": "@haipham/javascript-helper-peer",
  "publishConfig": {
    "access": "public"
  },
  "resolutions": {
    "@types/lodash-es": "4.17.4",
    "lodash-es": "4.17.20",
    "peerjs": "1.3.1",
    "rxjs": "6.6.3"
  },
  "scripts": {
    "build": "echo \"Skipped\"",
    "test": "jest"
  },
  "version": "1.0.1"
}
