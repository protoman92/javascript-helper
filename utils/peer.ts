import { cloneDeepWith } from "lodash-es";
import { PeerMessageCompatible } from "../interface";

const IDENTIFIER_KEY = "_peer_message_compatible_type";
const IDENTIFIER_VALUE_MAP = "map";
const IDENTIFIER_VALUE_SET = "set";

namespace TypeIdentifiable {
  export interface Map {
    [IDENTIFIER_KEY]: typeof IDENTIFIER_VALUE_MAP;
    [x: number]: unknown;
    [x: string]: unknown;
  }

  export interface Set {
    [IDENTIFIER_KEY]: typeof IDENTIFIER_VALUE_SET;
    [x: number]: unknown;
    length: number;
  }
}

type TypeIdentifiable = TypeIdentifiable.Map | TypeIdentifiable.Set;

export const serializePeerMessage = (() => {
  function customizer(value: unknown) {
    if (value instanceof Map) {
      const serialized: TypeIdentifiable.Map = {
        [IDENTIFIER_KEY]: IDENTIFIER_VALUE_MAP,
      };

      for (const [k, v] of value) serialized[k] = clone(v);
      return serialized;
    } else if (value instanceof Set) {
      const serialized: TypeIdentifiable.Set = {
        [IDENTIFIER_KEY]: IDENTIFIER_VALUE_SET,
        length: 0,
      };

      let currentIndex = 0;

      /**
       * To preserve the identifier key, we need to serialize the set into
       * an object with number keys, instead of an array, since the latter will
       * have its identifier key removed on message receipt.
       */
      for (const v of value) {
        serialized[currentIndex] = clone(v);
        currentIndex += 1;
      }

      serialized.length = currentIndex;
      return serialized;
    }

    return undefined;
  }

  function clone<T>(serializable: T): PeerMessageCompatible<T> {
    return cloneDeepWith(serializable, customizer);
  }

  return clone;
})();

export const deserializePeerMessage = (() => {
  function isTypeIdentifiable(value: unknown): value is TypeIdentifiable {
    return (
      typeof value === "object" && value != null && IDENTIFIER_KEY in value
    );
  }

  function isMapTypeIdentifiable(
    value: unknown
  ): value is TypeIdentifiable.Map {
    return (
      isTypeIdentifiable(value) &&
      value[IDENTIFIER_KEY] === IDENTIFIER_VALUE_MAP
    );
  }

  function isSetTypeIdentifiable(
    value: unknown
  ): value is TypeIdentifiable.Set {
    return (
      isTypeIdentifiable(value) &&
      value[IDENTIFIER_KEY] === IDENTIFIER_VALUE_SET
    );
  }

  function customizer(value: unknown) {
    if (isMapTypeIdentifiable(value)) {
      const deserialized = new Map();

      for (const k in value) {
        if (k === IDENTIFIER_KEY) continue;
        deserialized.set(k, clone(value[k]));
      }

      return deserialized;
    } else if (isSetTypeIdentifiable(value)) {
      const deserialized = new Set();
      const arrayLength = value.length;

      for (let index = 0; index < arrayLength; index += 1) {
        deserialized.add(clone(value[index]));
      }

      return deserialized;
    }

    return undefined;
  }

  function clone<T>(deserializable: PeerMessageCompatible<T>): T {
    return cloneDeepWith(deserializable, customizer);
  }

  return clone;
})();
