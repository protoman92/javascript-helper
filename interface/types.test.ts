import { NonNullableProps } from "./essentials";
import { I18N } from "./i18n";
import { PeerMessageCompatible } from "./peer";

describe("Essential types", () => {
  it("Non nullable props should work", async () => {
    // Setup
    type A = { a?: number; b?: { a?: number; b?: { a?: number } } };
    type B = NonNullableProps<A>;
    type C = NonNullableProps<A, "a">;

    // When
    const b: B = { a: 1, b: {} };
    const c: C = { a: 1 };

    // Then
    expect(b.b).toBeTruthy();
    expect(c.a).toBeTruthy();
  });
});

describe("i18n types", () => {
  it("Should have correct t methods", async () => {
    // Setup
    type LocaleContent = { a: { b: { c: string } }; b: { c: string } };
    type Client = I18N<LocaleContent>;

    // When
    const localeContent = { a: { b: { c: "1" } }, b: { c: "2" } };

    const client: Client = {
      t: (...keys: readonly (object | string)[]) => {
        let currentObject: I18N.LocaleContent = localeContent;

        if (
          keys.length === 1 &&
          typeof keys[0] === "string" &&
          keys[0].includes(".")
        ) {
          keys = keys[0].split(".");
        }

        let replacement: object | undefined;

        if (typeof keys[keys.length - 1] === "object") {
          replacement = keys[keys.length - 1] as object;
        }

        for (const key of keys) {
          if (typeof key === "object") {
            throw new Error("Wrong key");
          } else if (key in currentObject) {
            const childObject = currentObject[key];

            if (typeof childObject === "object") {
              currentObject = childObject;
            } else {
              return `${childObject}${JSON.stringify(replacement)}`;
            }
          } else {
            return `${key}${JSON.stringify(replacement)}`;
          }
        }

        throw new Error("Not found");
      },
    } as any;

    const replacement = { x: 1, y: 2, z: 3 };

    // Then
    expect(client.t("a", "b", "c", replacement)).toMatchSnapshot();
    expect(client.t("b", "c", replacement)).toMatchSnapshot();
    expect(client.t("a.b.c", replacement)).toMatchSnapshot();
    expect(client.t("whatever", replacement)).toMatchSnapshot();
  });
});

describe("Peer types", () => {
  it("Peer message compatible type should work", async () => {
    // Setup
    type A = Map<
      string,
      { a: Set<number>; b: { c: Set<string>; d: number[] } }
    >;

    const a: PeerMessageCompatible<A> = {
      a: { a: [1], b: { c: ["a"], d: [2] } },
    };

    // Then
    expect(a["a"].a).toBeInstanceOf(Array);
    expect(a["a"].b.c).toBeInstanceOf(Array);
    expect(a["a"].b.d).toBeInstanceOf(Array);
  });
});
