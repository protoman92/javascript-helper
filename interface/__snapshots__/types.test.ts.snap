// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`i18n types Should have correct t methods 1`] = `"1{\\"x\\":1,\\"y\\":2,\\"z\\":3}"`;

exports[`i18n types Should have correct t methods 2`] = `"2{\\"x\\":1,\\"y\\":2,\\"z\\":3}"`;

exports[`i18n types Should have correct t methods 3`] = `"a.b.c{\\"x\\":1,\\"y\\":2,\\"z\\":3}"`;

exports[`i18n types Should have correct t methods 4`] = `"whatever{\\"x\\":1,\\"y\\":2,\\"z\\":3}"`;
